lisa/sut_orchestrator/azure/tests/test_prepare.py

```diff
@@ -0,0 +1,331 @@
+        return result
+
+    def verify_eligible_vm_size(
+        self, location: str, vm_size: str, expect_exists: bool
+    ) -> Optional[azure.AzureCapability]:
+        result = None
+        assert self._platform._eligible_capabilities
+        self.assertEqual(
+            expect_exists,
+            any(
+                [
+                    x.vm_size == vm_size
+                    for x in self._platform._eligible_capabilities[location]
+                ]
+            ),
+        )
+        if expect_exists:
+            result = next(
+                x
+                for x in self._platform._eligible_capabilities[location]
+                if x.vm_size == vm_size
+            )
+        return result
+
+    def load_environment(self, node_req_count: int = 2,) -> Environment:
+        environment = Environment(is_predefined=True, warn_as_error=False)
+        environment.runbook = schema.Environment()
+        if node_req_count > 0:
+            environment.runbook.nodes_requirement = []
+
+            for _ in range(node_req_count):
+                node_req = schema.NodeSpace()
+                _ = node_req.get_extended_runbook(azure.AzureNodeSchema, azure.AZURE)
+                environment.runbook.nodes_requirement.append(node_req)
+
+        return environment
+
+    def set_node_runbook(
+        self,
+        environment: Environment,
+        index: int,
+        location: str = "",
+        vm_size: str = "",
+    ) -> None:
+        assert environment.runbook.nodes_requirement
+        node_runbook = environment.runbook.nodes_requirement[
+            index
+        ].get_extended_runbook(azure.AzureNodeSchema, azure.AZURE)
+        node_runbook.location = location
+        node_runbook.vm_size = vm_size
+
+    def verify_prepared_nodes(
+        self,
+        expected_result: bool,
+        expected_locations: List[str],
+        expected_vm_sizes: List[str],
+        expected_cost: int,
+        environment: Environment,
+    ) -> None:
+        actual_result = self._platform._prepare_environment(environment, self._log)
+        self.assertEqual(expected_result, actual_result)
+
+        if expected_locations:
+            assert environment.runbook.nodes_requirement
+
+            # get node runbook for validating
+            nodes_runbook = [
+                x.get_extended_runbook(azure.AzureNodeSchema, azure.AZURE)
+                for x in environment.runbook.nodes_requirement
+            ]
+
+            self.assertListEqual(
+                expected_locations, [x.location for x in nodes_runbook],
+            )
+            self.assertListEqual(
+                expected_vm_sizes, [x.vm_size for x in nodes_runbook],
+            )
+
+            # all cap values must be covered to specified int value, not space
+            for node_cap in environment.runbook.nodes_requirement:
+                assert node_cap
+                self.assertIsInstance(node_cap.core_count, int)
```

Comment:
Might be nice to also check that these numbers are sane -- for instance, `>= 1` CPUs, `>= 1` disks, `>= 1` NICs, etc.

Reply:
Also, might be a good idea to check if the quantity of Azure NICs is at least `1 <= number of NICs in the system`. You can check that by parsing `/proc/net/dev` (ignore the first line, and count the number of lines; that'll give you the number of interfaces in that system).  This will include things like`lo` for the loopback device, which isn't a NIC, but should be fine for these assertions here.

Reply:
Will add number checking to increase coverage. The assertions of isinstance, they will be kept here, as it's meaningful like "the core count must be an integer". The purpose of the assertions are make sure the capabilty, which type is nodeSpace, shouldn't be a range anymore, they should be a clear number now.

It's a check on a memory object, no actual node will be created here. So there is no check on actual system. For LISAv3 itself, there is no such validation yet. We can add it, if it's a common requirement.